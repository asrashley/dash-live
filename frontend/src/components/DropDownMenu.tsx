import type { ComponentChildren } from "preact";
import { useCallback, useState } from "preact/hooks";
import { MenuItemType } from "../types/MenuItemType";

export interface DropDownItemProps extends Omit<MenuItemType, "title"> {
  index: number;
  children: ComponentChildren;
  setExpanded: (value: boolean) => void;
}

function DropDownItem({
  href = "#",
  index,
  onClick,
  setExpanded,
  children,
}: DropDownItemProps) {
  const click = useCallback(
    (ev) => {
      ev.preventDefault();
      setExpanded(false);
      onClick(ev);
    },
    [onClick, setExpanded]
  );
  return (
    <li data-testid={`ddi_${index}`}>
      <a className="dropdown-item" href={href} onClick={click}>
        {children}
      </a>
    </li>
  );
}

export interface DropDownMenuProps {
  children?: ComponentChildren;
  linkClass?: string;
  menu: MenuItemType[];
}
export function DropDownMenu({
  children,
  menu,
  linkClass = "btn btn-secondary",
}: DropDownMenuProps) {
  const [expanded, setExpanded] = useState(false);

  const toggleMenu = (ev) => {
    ev.preventDefault();
    setExpanded(!expanded);
  };

  const show = expanded ? " show" : "";

  return (
    <div className="dropdown">
      <a
        class={`${linkClass} dropdown-toggle${show}`}
        href="#"
        onClick={toggleMenu}
        role="button"
        aria-expanded={expanded}
      >
        {children}
      </a>
      <ul className={`dropdown-menu ${show}`}>
        {menu.map((item, index) => (
          <DropDownItem
            key={item.title}
            index={index}
            onClick={item.onClick}
            setExpanded={setExpanded}
            href={item.href}
          >
            {item.title}
          </DropDownItem>
        ))}
      </ul>
    </div>
  );
}
