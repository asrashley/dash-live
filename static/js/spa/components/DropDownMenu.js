import { html } from 'htm/preact';
import { useCallback, useState } from 'preact/hooks';

function DropDownItem({href = "#", onClick, setExpanded, children}) {
    const click = useCallback((ev) => {
        ev.preventDefault();
        setExpanded(false);
        onClick(ev);
    }, [onClick, setExpanded]);
  return html`<li>
    <a class="dropdown-item" href="${href}" onClick=${click}>${children}</a>
    </li>`;
}

export function DropDownMenu({children, menu, linkClass="btn btn-secondary"}) {
    const [expanded, setExpanded] = useState(false);

    const toggleMenu = (ev) => {
        ev.preventDefault();
        setExpanded(!expanded);
    };

    const show = expanded ? 'show': '';

    return html`<div class="dropdown">
  <a class="${linkClass} dropdown-toggle ${show}" href="#" onClick=${toggleMenu}
    role="button" aria-expanded=${expanded}>
    ${children}
  </a>
  <ul class="dropdown-menu ${show}">
  ${menu.map(item => html`
    <${DropDownItem} onClick=${item.onClick} setExpanded=${setExpanded} href="${item.href}">${item.title}<//>`)}
  </ul>
</div>`;
}