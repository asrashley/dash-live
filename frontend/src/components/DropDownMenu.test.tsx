import { vi, describe, expect, test } from "vitest";
import { fireEvent } from "@testing-library/preact";

import { renderWithProviders } from "../test/renderWithProviders";
import { DropDownMenu} from './DropDownMenu';

describe('DropDownMenu', () => {
    test("should display drop-down menu", () => {
        const menu = [{
            href: '#',
            onClick: vi.fn(),
            title: 'item one',
        },
        {
            href: '/two',
            onClick: vi.fn(),
            title: 'item two',
        }
        ];
        const { asFragment, getByTestId, queryBySelector } = renderWithProviders(
          <DropDownMenu menu={menu} />
        );
        expect((getByTestId('ddi_0') as HTMLElement).textContent).toEqual('item one');
        expect((getByTestId('ddi_1') as HTMLElement).textContent).toEqual('item two');

        const elt = queryBySelector('.dropdown > a');
        expect(elt).not.toBeNull();
        expect(elt.className).toEqual('btn btn-secondary dropdown-toggle');
        expect(asFragment()).toMatchSnapshot();

        fireEvent.click(elt);
        expect(elt.className).toEqual('btn btn-secondary dropdown-toggle show');
        fireEvent.click((getByTestId('ddi_0') as HTMLElement).querySelector('a'));
        expect(menu[0].onClick).toHaveBeenCalledTimes(1);
        expect(menu[1].onClick).not.toHaveBeenCalled();
        expect(elt.className).toEqual('btn btn-secondary dropdown-toggle');

        fireEvent.click(elt);
        fireEvent.click((getByTestId('ddi_1') as HTMLElement).querySelector('a'));
        expect(menu[0].onClick).toHaveBeenCalledTimes(1);
        expect(menu[1].onClick).toHaveBeenCalledTimes(1);
      });

});