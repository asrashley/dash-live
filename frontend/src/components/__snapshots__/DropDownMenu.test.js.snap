// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDownMenu > should display drop-down menu 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <a
      aria-expanded="false"
      class="btn btn-secondary dropdown-toggle"
      href="#"
      role="button"
    />
    <ul
      class="dropdown-menu "
    >
      <li
        data-testid="ddi_0"
      >
        <a
          class="dropdown-item"
          href="#"
        >
          item one
        </a>
      </li>
      <li
        data-testid="ddi_1"
      >
        <a
          class="dropdown-item"
          href="/two"
        >
          item two
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
